core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -63,6 +65,8 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
         doWork()
       }
     } catch{
+      case fatalExitError: FatalExitError =>
+        fatalExitError.shutdownSystem()
```

Comment:
Thanks @resetius  For the benefit of future readers, here is the excerpt from the stack trace that shows the shutdown hook is waiting for the fetcher threads to finish, which one of them is the one who initiated System.exit.

```
    at kafka.server.AbstractFetcherManager.closeAllFetchers(AbstractFetcherManager.scala:122)
    - locked <0x00000003e8bdf750> (a java.lang.Object)
    at kafka.server.ReplicaFetcherManager.shutdown(ReplicaFetcherManager.scala:41)
    at kafka.server.ReplicaManager.shutdown(ReplicaManager.scala:901)
    at kafka.server.KafkaServer$$anonfun$shutdown$7.apply$mcV$sp(KafkaServer.scala:581)
    at kafka.utils.CoreUtils$.swallow(CoreUtils.scala:76)
    at kafka.utils.Logging$class.swallowWarn(Logging.scala:92)
    at kafka.utils.CoreUtils$.swallowWarn(CoreUtils.scala:47)
    at kafka.utils.Logging$class.swallow(Logging.scala:94)
    at kafka.utils.CoreUtils$.swallow(CoreUtils.scala:47)
    at kafka.server.KafkaServer.shutdown(KafkaServer.scala:581)
    at kafka.server.KafkaServerStartable.shutdown(KafkaServerStartable.scala:49)
```

So it seems that with the current implementation of the shutdown hook calling System.exit from any thread known to Kafka would cause a deadlock.

@resetius I still like your solution of invoking System.exit from an anonymous thread (daemon). In that case the invoking thread would hold no lock and yet being anonymous the shutdown hook would not try to join it. But it seems that you reported that it did not work. Can you provide more information of why?

> In fact, my "solution" from KAFKA-3924 does not solve the problem because shutdown hook waits for sheduler and vice versa.